R/class2tree.R

```diff
@@ -241,82 +249,98 @@ rank_indexing <- function(rankList){
+          if (j < i) {
+            if (!is.null(tmpEnv[[subList[i - j]]])) {
+              tmpEnv[[iRank]] <- tmpEnv[[subList[i - j]]] + 1
+              break
+            }
+          } else j = j - 1
         }
-      }
-      ## else, check if the current index is smaller than
-      ## the index of previous rank,
-      else {
-        if(i>1){
-          preRank <- subList[i-1]
-          ## if so, increase index of this current rank
-          ## by (index of previous rank + 1)
-          if(rank2Index[[subList[i]]] <= rank2Index[[preRank]]){
-            rank2Index[[subList[i]]] = rank2Index[[preRank]] + 1
+      } else {
+        # for old rank
+        if (i > 1) {
+          if (rank2index[[iRank]] < tmpEnv[[subList[i-1]]]) {
+            tmpEnv[[iRank]] <- tmpEnv[[subList[i-1]]] + 1
+            for (
+              r in 
+              ls(rank2index)[!(ls(rank2index) %in% ls(tmpEnv))]
+            ) {
+              if (rank2index[[r]] >= tmpEnv[[subList[i-1]]]) {
+                tmpEnv[[r]] <- 
+                  rank2index[[r]] + tmpEnv[[iRank]] - 
+                  rank2index[[iRank]]
+              }
+            }
+          } else {
+            if (is.null(tmpEnv[[iRank]]))
+              tmpEnv[[iRank]] <- rank2index[[iRank]]
           }
-        }
+        } else tmpEnv[[iRank]] <- rank2index[[iRank]]
       }
     }
+    assignHash(ls(tmpEnv), getHash(ls(tmpEnv), tmpEnv), rank2index)
   }
-
-  ### output a list of indexed ranks
-  index2RankDf <- data.frame("index"=character(),
-                             "rank"=character(),
-                             stringsAsFactors=FALSE)
-  for(i in 1:length(allInputRank)){
-    index2RankDf[i,] = c(rank2Index[[allInputRank[i]]],allInputRank[i])
-  }
-
-  index2RankDf$index <- as.numeric(index2RankDf$index)
+  
+  # convert env into dataframe and return
+  index2RankList <- lapply(
+    seq_len(length(allInputRank)), function (x) {
+      data.frame(
+        index = rank2index[[allInputRank[x]]],
+        rank = allInputRank[x], stringsAsFactors = FALSE
+      )
+    }
+  )
+  index2RankDf <- do.call(rbind, index2RankList)
   index2RankDf <- index2RankDf[with(index2RankDf, order(index2RankDf$index)),]
-
   return(index2RankDf)
 }
 
-###############################################################################
-#################### ARRANGE RANK IDs INTO SORTED RANK LIST (index2RankDf) ####
-###############################################################################
-taxonomy_table_creator <- function(nameList,rankList){
+#' Align NCBI taxonomy IDs of list of taxa into a sorted rank list
+#' @param nameList a dataframe whose each row is a rank+ID list of a taxon
+#' @param rankList a dataframe whose each row is a rank list of a taxon
+#' @return An aligned taxonomy dataframe which contains all the available
+#' taxonomy ranks from the id list and rank list
+#' @author Vinh Tran {tran@bio.uni-frankfurt.de}
+#' @import data.table
```

Comment:
we have a central place to list imports https://github.com/ropensci/taxize/blob/master/R/taxize-package.R#L69 - lets try to use importFrom instead of import (which imports the whole package)

Reply:
ok, I've removed it. I saw that the `data.table` already imported from the DESCRIPTION file.